chain/chain/src/chain.rs

```diff
@@ -1529,6 +1560,18 @@ impl Chain {
         ),
         Error,
     > {
+        // We set a limit to the max number of blocks that we will be processing at the same time.
+        // Since processing a block requires that the its previous block is processed, this limit
+        // is likely never hit, unless there are many forks in the chain.
+        // In this case, we will simply drop the block.
+        if self.processing_blocks.is_full() {
+            return Err(ErrorKind::TooManyProcessingBlocks.into());
+        }
+
+        if self.processing_blocks.is_block_preprocessed(block.hash()) {
```

Comment:
do we need this here? (or should we fail when adding the preprocessing info instead?)
when do we expect to trigger this?

Reply:
I think we should check before we actually start to preprocess the block because we can save time by failing early. Imagine if we receive the same block multiple times, this would save us time by not having to preprocessing the same block over and over.